### PR TITLE
[SPARK-23550][core] Cleanup `Utils`.

### PR DESCRIPTION
**Pull request by vanzin**
A few different things going on:
- Remove unused methods.
- Move JSON methods to the only class that uses them.
- Move test-only methods to TestUtils.
- Make getMaxResultSize() a config constant.
- Reuse functionality from existing libraries (JRE or JavaUtils) where possible.

The change also includes changes to a few tests to call `Utils.createTempFile` correctly,
so that temp dirs are created under the designated top-level temp dir instead of
potentially polluting git index.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -267,44 +264,20 @@ private[spark] object Utils extends Logging {
...
-        throw new IOException("Failed to create a temp directory (under " + root + ") after " +
-          maxAttempts + " attempts!")
-      }
-      try {
-        dir = new File(root, namePrefix + "-" + UUID.randomUUID.toString)
-        if (dir.exists() || !dir.mkdirs()) {
-          dir = null
-        }
-      } catch { case e: SecurityException => dir = null; }
+    val prefix = namePrefix + "-"
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: was there a reason you rewriting this?

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: This code was written when Spark still supported Java 1.6, and `java.nio.file` did not exist.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -267,44 +264,20 @@ private[spark] object Utils extends Logging {
...
+    val dir = if (root != null) {
+      val rootDir = new File(root)
+      require(rootDir.isDirectory(), s"Root path $root must be an existing directory.")
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Why not use the nio api's? e.g.:
```scala
val rootDir = Paths.get(root)
require(Files.isDirectory(rootDir), s"Root path $root must be an existing directory.")
Files.createTempDirectory(rootDir, prefix)
```

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Same thing in the end, no? This method needs to return `File` regardless of what it does internally.

Comment (hvanhovell) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: That is also an argument for not changing this method at all. Just use `File.createTempFile(...)` and be done with it.

**Comment by gatorsmile** [comment]
The behavior changes have to be separated from the code refactoring.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: This is for a directory, not file. `File.createTempFile` does not work.

**Comment by vanzin** [comment]
> The behavior changes have to be separated from the code refactoring.

The behavior change only affects tests, which have been fixed in this PR.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (hvanhovell) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Yeah, I missed that sorry. However my point still stands.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Not sure what that point really is, but if you feel so strongly about using the nio APIs, that's easy enough.

**Comment by gatorsmile** [comment]
Unfortunately, I do not know whether this change only affects the test cases. Any behavior change require a conf. We can get rid of it later if nobody hit the issue.

**Comment by vanzin** [comment]
>  I do not know whether this change only affects the test cases.

I do, because I checked. The tests were using the API incorrectly, actually, and could leave garbage outside the temp dirs, polluting the source tree. That wrong usage pattern was actually the thing that made tests be affected by the change. No code outside of tests uses the API in that way.

And adding a conf would kinda defeat the purpose of using the existing JRE API, since it would require maintaining the old conf.

In summary, not all behavior changes are equal, especially if you're talking about an internal API.

That being said, I'm going to run some tests with YARN+Kerneros+ShuffleService to make sure things are the same - I think I may need to adjust something in `DiskBlockManager`.

**Comment by vanzin** [comment]
Ok, I'm going to revert the `createDirectory` changes because the `Files.createTempDirectory` api forces permissions to `700`, and secure YARN requires that the block manager directories respect the umask set for the container or things break. I could fix that in `DiskBlockManager` but reverting sounds safer.

I'll keep the test changes since they're correct and fix a potential issue with garbage being left around the working dir.

**Comment by vanzin** [comment]
I'm almost sure there's a bug somewhere for that flaky test, but if there isn't I'll file one.

**Comment by vanzin** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -1018,70 +1016,18 @@ private[spark] object Utils extends Logging {
...
-      files
-    } else {
-      List()
-    }
-  }
-
-  /**
-   * Lists files recursively.
-   */
-  def recursiveList(f: File): Array[File] = {
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Yes this is only used by test suites for now, but can you still keep this in `Utils` so we don't have to move it back when we want to use it in the production code (I assume this may happen someday).

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 4]
```diff
@@ -1876,17 +1822,6 @@ private[spark] object Utils extends Logging {
     obj.getClass.getSimpleName.replace("$", "")
   }
 
-  /** Return an option that translates JNothing to None */
-  def jsonOption(json: JValue): Option[JValue] = {
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: This is only used in `JSONProtocol` for now, but this is also a useful JSON method that could apply to some other JSON operations in the future.

**Review by jiangxb1987 — COMMENTED** [review]
I looked at some of the changes that move methods out of `Utils`, they seems valid till now, but I'm not sure whether these are good move because I'm afraid we may have to use them somewhere else in the future and then we have to move them back to `Utils`. I really feel we shall be conservative to make these changes.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 4 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: We should not be using json4s for new things, so it's better to have these hidden away.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 3 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: I disagree that we should keep things in production code "just in case".

This is also an example of a function that if used without care is dangerous. `Files.walk` is a much better way of achieving this.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 4 above)
Comment (gatorsmile) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: What is the reason we should not use json4s for new things?

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: It encourages writing manual serialization and deserialization code, which is error prone.

**Comment by hvanhovell** [comment]
@jiangxb1987 keeping dead code around will inevitably lead to code rot. If there is not use for something then we should remove it. Adding things back is not a problem.

**Comment by vanzin** [comment]
Any more comments?

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master.
